php-checks/src/main/java/org/sonar/php/checks/regex/GroupReplacementCheck.java

```diff
@@ -40,26 +40,19 @@
   private static final String MESSAGE = "Referencing non-existing group%s: %s.";
   private static final Pattern REFERENCE_PATTERN = Pattern.compile("\\$(\\d+)|\\$\\{(\\d+)}|\\\\(\\d+)");
 
-  private final Set<CapturingGroupTree> groups = new HashSet<>();
-
   @Override
   protected Set<String> lookedUpFunctionNames() {
     return Set.of("preg_replace");
   }
 
-  @Override
-  protected void checkFunctionCall(FunctionCallTree tree) {
-    super.checkFunctionCall(tree);
-    checkReplacement(tree);
-  }
-
   @Override
   public void checkRegex(RegexParseResult regexParseResult, FunctionCallTree regexFunctionCall) {
-    groups.clear();
-    new GroupFinder().visit(regexParseResult);
+    GroupFinder groupFinder = new GroupFinder();
+    groupFinder.visit(regexParseResult);
+    checkReplacement(regexFunctionCall, groupFinder.groups);
   }
 
-  private void checkReplacement(FunctionCallTree tree) {
+  private void checkReplacement(FunctionCallTree tree, Set<CapturingGroupTree> groups) {
```

Comment:
I don't get it - why add this second parameter 'groups' when you don't seem to use it?

Reply:
It's not a field anymore. It has to be passed as a paramerter. So, the Set does not need to be cleared after every check.